### ISSUE TITLE
Unattempted answers in a quiz should not show up as difficult questions?

### ISSUE THREAD

**Issue by radinamatic** [learningequality/kolibri#5626]
### Observed behavior
Unattempted answers in a quiz show up as _difficult questions_  immediately upon learner opening the quiz.

![unattempted-diffque](https://user-images.githubusercontent.com/1457929/58970422-b2349200-87b9-11e9-9231-e55c60880406.gif)

In the above example all 5 questions are categorized as difficult even though the second learner has just started the quiz, and haven't even viewed the questions 2-5.

### Expected behavior
Label questions as difficult _only_ after the quiz has been submitted? Or more complex one would be to have the `viewed and skipped more than once` as the threshold for unattempted question labeled as difficult?

### User-facing consequences
Coaches getting overestimates about the number of difficult questions. 

### Errors and logs

…

### Steps to reproduce

…

### Context

 * Kolibri version: 0.12.4b6 upgraded from 0.10
 * Operating system: Ubuntu 16.04
 * Browser: all

cc @rtibbles @indirectlylit (I did not put the milestone, in case the fix is out of the scope for now)

**Comment by jonboiser** [comment]
The current implementation of `showQuestionListView` and the `setItemStats` action intentionally shows all of the exam questions, even if no attempts have been logged for some questions.

So, is the fix here is to change that implementation to show a link to a question only if it has been flagged as "difficult" via some logged activity?

**Comment by rtibbles** [comment]
I think the fix is that the backend should only calculate the statistics for this based on user's attempts on quizzes for which the quiz is closed or the examlog is marked as completed.

**Comment by jonboiser** [comment]
@rtibbles Is this how to change the backend query to match your description of the new requirement:

From 

https://github.com/learningequality/kolibri/blob/554b27f3c457504dc86398543b8bdbafcc5af11c/kolibri/plugins/coach/api.py#L331

```diff
-         queryset = ExamAttemptLog.objects.filter(examlog__exam=pk)
# examlog__closed=True -> Learner Submitted quiz
# examlog__exam__active=False -> Coach changed quiz status to inactive
+        queryset = ExamAttemptLog.objects.filter(Q(examlog__closed=True) | Q(examlog__exam__active=False), examlog__exam=pk)

```

**Comment by rtibbles** [comment]
Yeah, looks exactly right to me.

**Comment by jonboiser** [comment]
There's also some extra front-end code that will fill in gaps from the Difficulties ExamAttemptLogs. I think this should be removed as well. So a link to a difficult question should only appear if it is returned from the QuizDifficultiesViewset.

So in the snippet below, we will not add a `stat` array element for a quiz question if it's not returned from the backend.

https://github.com/learningequality/kolibri/blob/554b27f3c457504dc86398543b8bdbafcc5af11c/kolibri/plugins/coach/assets/src/modules/questionList/actions.js#L48-L63

**Comment by rtibbles** [comment]
Hrm - would still be good to keep those though - if there are questions that were completely unattempted by everyone who took a quiz, then we should flag that because it is a problematic question for some reason.

**Comment by jonboiser** [comment]
That's a good point. I think this front-end code needs to implement a similar rule as the backend then:

> Show Questions with no Attempts only if the quiz is closed

(By definition, these Questions will still not be returned from the backend because they have no ExamAttemptLogs, so they will also not appear while the quiz is still active/being taken by learners)

**Comment by rtibbles** [comment]
Does the current code not achieve that - if there are no users who have finished the quiz, then:
`total: (stats[0] || {}).total || 0,`
will evaluate to zero - and I think we exclude questions with total of zero (and if we don't,we should!)

**Comment by jonboiser** [comment]
I don’t think that’s how it works currently. As soon as a single learner starts a quiz, and either skips or gets questions wrong, all of the questions will be listed as “difficult”.

The stats appear because “total” will be non-zero as soon as a single difficulty is logged and, with the modified queryset, the exam is completed by a least one person.

I think I'm confusing myself as to what the bar is for a question to be difficult on a quiz. Right now it seems to be:

1. At least one quiz taker got it wrong (in-progress quiz-takers who skip/don't get to it it get lumped in the report too)
1. At least one quiz taker skippped it (but in current implementation even in-progress quiz takers get lumped in with finished quiz-takers who skipped it).

https://github.com/learningequality/kolibri/blob/554b27f3c457504dc86398543b8bdbafcc5af11c/kolibri/plugins/coach/api.py#L344-L353

**Comment by jonboiser** [comment]
Here's how I think it should work in different scenarios:

1. Nobody has started the quiz -> No difficult questions
1. Everybody has completed the quiz and got everything right -> No difficult questions
1. N quiz-takers got a question wrong/skipped question, but none of them finished quiz, and quiz is still active -> No difficult questions
1. N quiz-takers got a question wrong/skipped question, subset of M finished the quiz, and quiz is still active -> Question appears on list, with "(M of Total)" in "Help Needed" column. Only the M appears in the report when the link is clicked.
1. N quiz-takers got a question wrong/skipped question, and coach deactivated quiz -> Question appears on list with ("N of total") in "Help needed" column. All N quiz-takers appear in the report when the link is clicked.

**Comment by rtibbles** [comment]
That accords with my understanding also.

To add - the current threshold for showing a difficult question is that fewer than 80% of got it correct/completed it. This is completely arbitrary, but does give the slightly weird result that if only one user has finished a quiz, a question is considered difficult if that user got it wrong.

**Comment by rtibbles** [comment]
Also - I meant was that with the change to the backend, then the frontend logic should work as is, because people who have not completed the quiz (either by submitting or having it closed on them) will not be counted in the total.

**Comment by jonboiser** [comment]
Okay, I think I'm considering more nuanced scenarios, where `total` should vary depending on the specific question and whether the quiz is active or not and other factors. But it seems that the one-line change to the query set should fix the original issue 

> Unattempted answers in a quiz show up as difficult questions immediately upon learner opening the quiz.

Since a question will only appear on the "Difficult Questions" tab only when at least one user has submitted a quiz, or the quiz has been de-activated.

I still think there are remaining subtle issues with how the numbers are calculated in the "Needs Help" column, and who appears in the difficult question report (https://github.com/learningequality/kolibri/issues/5626#issuecomment-502855385)

**Comment by rtibbles** [comment]
Oh yes - sorry, we should also change the backend calculation of total too.

```diff
 total = ( 
-     HierarchyRelationsFilter(ExamLog.objects.filter(exam_id=pk))
+    HierarchyRelationsFilter(ExamLog.objects.filter(Q(closed=True) | Q(exam__active=False), exam_id=pk))
     .filter_by_hierarchy( 
         ancestor_collection=collection_id, target_user=F("user") 
     ) 
     .count() 
 ) 
```

**Comment by jonboiser** [comment]
Okay, so that change will make `total` equal to

1. If Quiz is active: The number of people who submitted the Quiz
1. If Quiz is inactive: The number of people who started the Quiz (regardless of whether they submitted it).

**Comment by rtibbles** [comment]
Yes - and I think that will then make the stats match the scenarios you outlined above.

**Comment by rtibbles** [comment]
Fixed in #5674
